AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.java

```diff
@@ -114,4 +119,121 @@ public void test_gendeck() {
+
+        models.save(model);
+        models.setCurrent(model);
+
+        Note note = col.newNote();
+        note.setItem("A", "foo");
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0, 1});
+
+        note = col.newNote();
+        note.setItem("B", "foo");
+        note.setItem("C", "foo");
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0, 1});
+
+        note = col.newNote();
+        note.setItem("B", "foo");
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0});
+
+        note = col.newNote();
+        note.setItem("C", "foo");
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0});
+
+        note = col.newNote();
+        note.setItem("A", "foo");
+        note.setItem("B", "foo");
+        note.setItem("C", "foo");
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0, 1});
+
+        note = col.newNote();
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0});
+        // First card is generated if no other card
+    }
+
+    @Test
+    public void test_gen_not() throws ConfirmModSchemaException {
+        Collection col = getCol();
+        Models models = col.getModels();
+        Model model = models.byName("Basic");
+        JSONArray flds = model.getJSONArray("flds");
+        JSONArray tmpls = model.getJSONArray("tmpls");
+
+        models.renameField(model, flds.getJSONObject(0), "First");
+        models.renameField(model, flds.getJSONObject(1), "Front");
+        JSONObject fld2 = models.newField("AddIfEmpty");
+        fld2.put("name", "AddIfEmpty");
+        models.addField(model, fld2);
+    
+        // ensure first card is always generated,
+        // because at last one card is generated
+        tmpls.getJSONObject(0).put("qfmt", "{{AddIfEmpty}}{{Front}}{{First}}");
+        JSONObject tmpl = models.newTemplate("NOT");
+        tmpl.put("qfmt", "    {{^AddIfEmpty}}        {{Front}}    {{/AddIfEmpty}}    ");
+    
+        models.addTemplate(model, tmpl);
+    
+        models.save(model);
+        models.setCurrent(model);
+    
+        Note note = col.newNote();
+        note.setItem("First", "foo");
+        note.setItem("AddIfEmpty", "foo");
+        note.setItem("Front", "foo");
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0});
+    
+        note = col.newNote();
+        note.setItem("First", "foo");
+        note.setItem("AddIfEmpty", "foo");
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0});
+    
+        note = col.newNote();
+        note.setItem("First", "foo"); // ensure first note generated
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0});
+    
+        note = col.newNote();
+        note.setItem("First", "foo");
+        note.setItem("Front", "foo");
+        col.addNote(note);
+        assert_ord(note, new Integer[]{0, 1});
+    }
+
+    private void  assert_ord(Note note, Integer[] ords) {
```

Comment:
```suggestion
    private void  assertNoteContainsCardOrdinals(Note note, Integer[] ords) {
```

Changes the really cryptic python-y name to something that means what it does so I don't have to read the method every time

Reply:
"assertNoteOrdinalAre" 
because I want to assert that this is the exact set of cards generated, and not only that those ords are contained

Reply:
I stuck a bit on the wording also, I think you picked well, it's even more accurate